modal/_utils/async_utils.py

```diff
@@ -544,3 +545,116 @@ async def producer(producer_id: int, iterable: Union[AsyncIterable[T], Iterable[
 
 async def callable_to_agen(awaitable: Callable[[], Awaitable[T]]) -> AsyncGenerator[T, None]:
     yield await awaitable()
+
+
+async def async_map(
+    input: Union[AsyncIterable[T], Iterable[T]],
+    async_mapper_func: Callable[[T], Awaitable[V]],
+    concurrency: int,
+) -> AsyncGenerator[V, None]:
+    input_queue: asyncio.Queue[Tuple[str, Union[T, None]]] = asyncio.Queue()
+    output_queue: asyncio.Queue[Tuple[str, Union[V, Exception]]] = asyncio.Queue()
+    output_event = asyncio.Event()
+
+    async def producer():
+        async for item in sync_or_async_iter(input):
+            await input_queue.put(("value", item))
+        await input_queue.put(("stop", None))
+
+    async def worker():
+        while True:
+            try:
+                event_type, item = await input_queue.get()
+                if event_type == "stop":
+                    break
+                if event_type == "value":
+                    if asyncio.iscoroutinefunction(async_mapper_func):
+                        result = await async_mapper_func(item)
+                    else:
+                        result = typing.cast(V, async_mapper_func(item))
+                    await output_queue.put(("value", result))
+            except Exception as e:
+                await output_queue.put(("exception", e))
+            finally:
+                input_queue.task_done()
+
+    output_event_task = asyncio.create_task(output_event.wait())
```

Comment:
I think this could be the source of a bug. `output_event_task` needs to be recreated every time that the output_event triggers, since clearing the event doesn't undo the completeness of `output_event_task`.

Otherwise the `asyncio.wait` here fires immediately in every loop iteration after the first input is returned, leading to a "hot loop" where the `output_event.is_set():` condition would be repeatedly checked even though nothing has changed.